lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {DeepRecord, Key, CollectionKey, Selector, Value} from './types';
+
+type KeyValueMap = {
+    [TKey in Key | CollectionKey]: Value[TKey];
+};
+
+type BaseConnectOptions<TKey extends Key | CollectionKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, Value[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends Key | CollectionKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, Value[TKey]> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: Value[TKey] | null, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+type MergeCollection<TKey extends Key, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+type UpdateOperation<TKey extends Key | CollectionKey> =
+    | {
+          onyxMethod: typeof METHOD.SET;
+          key: TKey;
+          value: Value[TKey];
+      }
+    | {
+          onyxMethod: typeof METHOD.MERGE;
+          key: TKey;
+          value: PartialDeep<Value[TKey]>;
+      }
+    | {
+          [TKey in CollectionKey]: {
+              onyxMethod: typeof METHOD.MERGE_COLLECTION;
+              key: TKey;
+              value: Record<`${TKey}${string}`, PartialDeep<Value[TKey]>>;
+          };
+      }[CollectionKey];
+
+type UpdateOperations<TKeyList extends (Key | CollectionKey)[]> = {
```

Comment:
Building off my suggestion above, I would prefer to call this OnyxUpdates.

Also, why can't this simply be an array of UpdateOperations?
```ts
type UpdateOperations = Array<UpdateOperation>;
```
Ideally
```ts
type OnyxUpdates = Array<OnyxUpdate>;
```
In that case maybe we don't need this separate type and we can avoid a plural type name.

Reply:
@neil-marcellini good question!

For each update operation, we have to figure out which key we're updating. Depending on the key, the value you can pass to `value` field of `OnyxUpdate` is different.

```ts
key: TKey;
value: Value[TKey];
```

With `type OnyxUpdates = Array<OnyxUpdate>;`, `OnyxUpdate` doesn't know anything about the key for which the update is done cannot infer type for the value field.

But with the following code, a correct type is inferred for each key in updates.

```ts
[TKey in keyof TKeyList]: UpdateOperation<TKeyList[TKey]>;
```

Reply:
@neil-marcellini Will rename to `OnyxUpdates` and `OnyxUpdate`.